Api integration

huzaifaumer01 opened the pull request:
Add Repository layer in architecture.

huzaifa-umer committed:
fruites api is integrated

huzaifa-umer committed:
repository implemented

UmairAfzalGitHub commented on Viper Practice/src/controllers/Cells/TaskTableViewCell.swift:
```diff
@@ -26,8 +26,8 @@ class TaskTableViewCell: UITableViewCell {
         // Configure the view for the selected state
     }
     
-    func configureCell(task: Task){
```
Class name is still TaskTableViewCell, should it be renamed as well?

huzaifaumer01 replied:
all classes names renamed from Task to Fruit

UmairAfzalGitHub commented on Viper Practice/src/controllers/TasksViewController.swift:
```diff
@@ -11,7 +11,7 @@ class TasksViewController: UIViewController {
     @IBOutlet weak var tableView: UITableView!
     var presenter : TaskPresenterProtocol?
-    var tasks = [Task]()
+    var tasks = [Fruit]()
```
same renaming for this class as well.
Also, why are you holding data in the viewController?

Our presenter should hold the data

huzaifaumer01 replied:
create a private var fruites in presenter and add getFruites func of return type array of fruites in presenter protocol and then use it in view controller.
let me know if this approach is okay or i have to create object of presenter class in view controller to get fruites data?

UmairAfzalGitHub commented on Viper Practice/src/entity/Fruit.swift:
```diff
@@ -0,0 +1,44 @@
+//  Fruit.swift
+//  Viper Practice
+//
+//  Created by ahsan shabir on 02/04/2024.
```
who is ahsan shabir?

huzaifaumer01 replied:
its one of Mac books account name

huzaifaumer01 replied:
i have changed the names to Huzaifa Umer

UmairAfzalGitHub commented on Viper Practice/src/entity/Fruit.swift:
```diff
@@ -0,0 +1,44 @@
+  var genus: String? = nil
+  var nutritions: Nutritions? = Nutritions()
+
+  enum CodingKeys: String, CodingKey {
```
I think we don't need coding keys if key name is same

huzaifaumer01 replied:
coding keys removed

UmairAfzalGitHub commented on Viper Practice/src/entity/Fruit.swift:
```diff
@@ -0,0 +1,44 @@
+  var family: String? = nil
+  var order: String? = nil
+  var genus: String? = nil
+  var nutritions: Nutritions? = Nutritions()
```
why is it optinal and at the same time assigning a default constructor?

huzaifaumer01 replied:
default constructor removed

UmairAfzalGitHub commented on Viper Practice/src/entity/Nutritions.swift:
```diff
@@ -0,0 +1,42 @@
+
+struct Nutritions: Codable {
+
+  var calories      : Float? = nil
```
why the extra space before the : ?

huzaifaumer01 replied:
extra spaces removed from all models

UmairAfzalGitHub commented on Viper Practice/src/entity/Nutritions.swift:
```diff
@@ -0,0 +1,42 @@
+  var carbohydrates : Float? = nil
+  var protein       : Float? = nil
+
+  enum CodingKeys: String, CodingKey {
```
same comment for these coding keys as well

huzaifaumer01 replied:
coding keys removed

UmairAfzalGitHub commented on Viper Practice/src/helpers/Localization/Strings.swift:
```diff
@@ -0,0 +1,180 @@
+    
+    struct Button {
+
+        static var settings: String {
```
we can reduce the extra number of lines by following:
 static let settings = "Settings"
 
 Also, we don't need var here

huzaifaumer01 replied:
removed extra string

UmairAfzalGitHub commented on Viper Practice/src/helpers/Localization/Strings.swift:
```diff
@@ -0,0 +1,180 @@
+        static let latestAppVersion = "https://gist.github.com/kinetic-solutions/bf5a7a73dca701fdc2af49ced00750e6/raw/PrayerTime-Version.json"
+        static let appStore = "itms-apps://apple.com/app/id1633114818"
+    }
+
```
extra empty lines

huzaifaumer01 replied:
removed extra strings

UmairAfzalGitHub commented on Viper Practice/src/networking/API.swift:
```diff
@@ -0,0 +1,40 @@
+
+import Foundation
+
+class API {
```
why not make it a protocol?

huzaifaumer01 replied:
create APIProtocol and add getFruites function in the protocol

UmairAfzalGitHub commented on Viper Practice/src/networking/API.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+}
+
+extension API {
```
there should be a separate class for FruitesApi

huzaifaumer01 replied:
create separate class FruitesApi and make API class generic

huzaifa-umer committed:
all classes renamed from Task to Fruit

huzaifa-umer committed:
hold fruites data in presenter

create a private var fruites in presenter and add getFruites func of return type array of fruites in presenter protocol and then use it in view controller.

huzaifa-umer committed:
remove extra spaces from model classes and remove Nutritions default constructor

huzaifa-umer committed:
removed extra strings

huzaifa-umer committed:
create separate FruitesApi class and and make API generic class

UmairAfzalGitHub commented on Viper Practice/src/controllers/FruitesViewController.swift:
```diff
@@ -1,17 +1,16 @@
 import UIKit
 
-class TasksViewController: UIViewController {
+class FruitesViewController: UIViewController {
```
make the class final

huzaifaumer01 replied:
class access modifier is changed to final

UmairAfzalGitHub commented on Viper Practice/src/controllers/FruitesViewController.swift:
```diff
@@ -22,23 +21,23 @@ class TasksViewController: UIViewController {
-extension TasksViewController : TaskViewProtocol {
-    func displayTasks(_ tasks: [Fruit]) {
-        self.tasks.append(contentsOf: tasks)
+extension FruitesViewController : FruitesViewProtocol {
```
we need to avoid extensions whel possible, add this protocol conformance to the original class

huzaifaumer01 replied:
fixed

UmairAfzalGitHub commented on Viper Practice/src/entity/Nutritions.swift:
```diff
@@ -2,41 +2,30 @@
-    case protein       = "protein"
-  
-  }
+  var calories: Float? = nil
```
we can remove default nil value

huzaifaumer01 replied:
fixed

UmairAfzalGitHub commented on Viper Practice/src/interactor/FruitesInteractor.swift:
```diff
@@ -0,0 +1,29 @@
+    var presenter: FruitesPresenterProtocol?
+    
+    func fetchFruites() {
+        let repo = FruitesRepository()
```
move this to class level

huzaifaumer01 replied:
fixed

UmairAfzalGitHub commented on Viper Practice/src/networking/FruitesApi.swift:
```diff
@@ -0,0 +1,26 @@
+//  FruitesApi.swift
+//  Viper Practice
+//
+//  Created by ahsan shabir on 21/04/2024.
```
name

huzaifaumer01 replied:
fixed

UmairAfzalGitHub commented on Viper Practice/src/presenter/FruitesPresenter.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    private var fruites = [Fruit]()
+    
+    func loadView() {
```
can be renamed to:
loadData()
fetchFruites()

huzaifaumer01 replied:
fixed

UmairAfzalGitHub commented on Viper Practice/src/presenter/FruitesPresenter.swift:
```diff
@@ -0,0 +1,36 @@
+protocol FruitesPresenterProtocol : AnyObject {
+    func loadView()
+    func fruitesFetched(_ fruits: [Fruit])
+    func getFruites() -> [Fruit]
```
convert it into a property

huzaifaumer01 replied:
fixed

UmairAfzalGitHub commented on Viper Practice/src/presenter/FruitesPresenter.swift:
```diff
@@ -0,0 +1,36 @@
+    func getFruites() -> [Fruit]
+}
+
+class FruitesPresenter : FruitesPresenterProtocol {
```
use MARK: -  for each protocol confromance

huzaifaumer01 replied:
MARK added in all classes

huzaifa-umer committed:
pull request feedback issues are resolved

UmairAfzalGitHub reviewed (approved):
Looks Good To Me
